File: 396_Rotate_Function/maxRotateFunction.py
class Solution(object):
    def rotate(self, A, p):
        for i in xrange(p):
            A[i], A[-1 - i] = A[-1 - i], A[i]

    def maxRotateFunction(self, A):
        n = len(A)
        res = 0
        for i in xrange(n):
            res = max(res, sum([idx * item for idx, item in enumerate(A)]))
            self.rotate(A, i + 1)
        return res

sol = Solution()
A = [4, 3, 2, 6]

print sol.maxRotateFunction(A)

File: 396_Rotate_Function/maxRotateFunction.cpp
#include <iostream>
#include <vector>
#include <numeric>

using namespace std;

class Solution {
    public:
        int maxRotateFunction(vector<int>& A) {
            int n = A.size();
            if(n == 0) return 0;
            int sum = 0;
            for(int n:A) sum += n;
            vector<int> F(n, 0);

            // calculate F[0]
            for(int i = 0; i < n; i++)
                F[0] += i * A[i];

            int res = F[0];
            for(int k = 1; k < n; k++) {
                F[k] = F[k - 1] + sum - A[n - k] * n;
                res = max(res, F[k]);
            }

            return res;
        }
};

int main() {
    return 0;
}
